providers/salesforce/bulk-write_test.go

```diff
@@ -0,0 +1,254 @@
+import (
+	"context"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"testing"
+
+	"github.com/amp-labs/connectors/common"
+	"github.com/amp-labs/connectors/common/interpreter"
+	"github.com/amp-labs/connectors/test/utils/mockutils"
+	"github.com/amp-labs/connectors/test/utils/mockutils/mockserver"
+	"github.com/amp-labs/connectors/test/utils/testroutines"
+	"github.com/amp-labs/connectors/test/utils/testutils"
+)
+
+func TestBulkWrite(t *testing.T) { //nolint:funlen,gocognit,cyclop,maintidx
+	t.Parallel()
+
+	responseCreateJob := testutils.DataFromFile(t, "bulk/write/launch-job-opportunity.json")
+	responseUpdateJob := testutils.DataFromFile(t, "bulk/write/update-job-opportunity.json")
+
+	bodyRequest := `{
+		"contentType":"CSV",
+		"externalIdFieldName":"external_id__c",
+		"lineEnding":"LF",
+		"object":"Opportunity",
+		"operation":"upsert"
+	}`
+
+	tests := []bulkWriteTestCase{
+		{
+			Name: "Mime response header expected",
+			Input: BulkOperationParams{
+				ObjectName:      "Opportunity",
+				ExternalIdField: "fieldName8",
+				CSVData:         strings.NewReader(""),
+				Mode:            UpsertMode,
+			},
+			Server:       mockserver.Dummy(),
+			ExpectedErrs: []error{interpreter.ErrMissingContentType},
+		},
+		{
+			Name: "Read object must be included",
+			Input: BulkOperationParams{
+				Mode: UpsertMode,
+			},
+			Server: httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				w.Header().Set("Content-Type", "application/json")
+				w.WriteHeader(http.StatusNoContent)
+			})),
+			ExpectedErrs: []error{common.ErrMissingObjects},
+		},
+		{
+			Name: "Delete mode cannot be used in BulkWrite",
+			Input: BulkOperationParams{
+				ObjectName: "Opportunity",
+				Mode:       DeleteMode,
+			},
+			Server:       mockserver.Dummy(),
+			ExpectedErrs: []error{ErrUnsupportedMode},
+		},
+		{
+			Name: "Upsert requires External ID",
+			Input: BulkOperationParams{
+				ObjectName: "Opportunity",
+				Mode:       UpsertMode,
+			},
+			Server:       mockserver.Dummy(),
+			ExpectedErrs: []error{ErrExternalIdEmpty},
+		},
+		{
+			Name: "CSV is required for upload",
+			Input: BulkOperationParams{
+				ObjectName:      "Opportunity",
+				ExternalIdField: "external_id__c",
+				CSVData:         nil,
+				Mode:            UpsertMode,
+			},
+			Server:       mockserver.Dummy(),
+			ExpectedErrs: []error{common.ErrMissingCSVData},
+		},
+		{
+			Name: "Creating Job fails on bad response",
+			Input: BulkOperationParams{
+				ObjectName:      "Opportunity",
+				ExternalIdField: "external_id__c",
+				CSVData:         strings.NewReader(""),
+				Mode:            UpsertMode,
+			},
+			Server: httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				w.Header().Set("Content-Type", "application/json")
+				w.WriteHeader(http.StatusInternalServerError)
+				_, _ = w.Write([]byte{})
+			})),
+			ExpectedErrs: []error{ErrCreateJob},
+		},
+		{
+			Name: "Create job id must be string",
```

Comment:
I am trying to understand, what is the significance of testing the responses from Salesforce by returning mock responses?

Is this to detect salesforce API change? We do not govern what kind of responses we get from Salesforce. So whether we get "Open" as state or not is not something we can test on. Just trying to understand reason behind this.

Reply:
The whole reason I introduced this mock tests which are now present in almost every deep connector is to capture the functionallity inside code. Otherwise, any change would require manually running all scripts. This is not always possible due to license expiration, even connector setup (ex: Salesforce needs custom fields to be declared for bulk tests to run), and it would be time consuming. Therefore, these unit tests imitate what provider would respond and enforce certain output, connector behaviour. You can think of it as regression testing.